src/flamegraph.cpp

```diff
@@ -653,65 +676,144 @@ FlameGraph::FlameGraph(QWidget* parent, Qt::WindowFlags flags)
+                setColorScheme(comboBox->currentData().value<Settings::ColorScheme>());
+            });
 
-    auto costAggregation = new QComboBox(this);
-    ResultsUtil::setupResultsAggregation(costAggregation);
-    auto costAggregationLabel = new QLabel(tr("Aggregate cost by:"));
-    costAggregationLabel->setBuddy(costAggregation);
-
-    m_searchInput = new QLineEdit(this);
-    m_searchInput->setPlaceholderText(i18n("Search..."));
-    m_searchInput->setToolTip(i18n("<qt>Search the flame graph for a symbol.</qt>"));
-    m_searchInput->setClearButtonEnabled(true);
-    connect(m_searchInput, &QLineEdit::textChanged, this, &FlameGraph::setSearchValue);
-    connect(this, &FlameGraph::uiResetRequested, this, [this]() { m_searchInput->clear(); });
-
-    auto controls = new QWidget(this);
-    controls->setLayout(new QHBoxLayout);
-    controls->layout()->setContentsMargins(0, 0, 0, 0);
-    controls->layout()->addWidget(m_backButton);
-    controls->layout()->addWidget(m_forwardButton);
-    controls->layout()->addWidget(m_costSource);
-    controls->layout()->addWidget(bottomUpCheckbox);
-    controls->layout()->addWidget(collapseRecursionCheckbox);
-    controls->layout()->addWidget(costThreshold);
-    controls->layout()->addWidget(costAggregationLabel);
-    controls->layout()->addWidget(costAggregation);
-    controls->layout()->addWidget(m_searchInput);
-    controls->layout()->addWidget(m_colorSchemeLabel);
-    controls->layout()->addWidget(m_colorSchemeSelector);
+            connect(Settings::instance(), &Settings::pathsChanged, this, [setColorScheme] {
+                if (Settings::instance()->colorScheme() == Settings::ColorScheme::System) {
+                    // setColorScheme triggers a redraw
+                    // we only need to redraw the flamegraph if the current color scheme is affected by the changed
+                    // paths
+                    setColorScheme(Settings::ColorScheme::System);
+                }
+            });
+        },
+        this);
+
+    auto searchInput = new CustomWidgetAction(
+        [this](QWidget* widget, QHBoxLayout* layout) {
+            auto searchInput = new QLineEdit(widget);
+            layout->addWidget(searchInput);
+
+            searchInput->setPlaceholderText(i18n("Search..."));
+            searchInput->setToolTip(i18n("<qt>Search the flame graph for a symbol.</qt>"));
+            searchInput->setClearButtonEnabled(true);
+            connect(searchInput, &QLineEdit::textChanged, this, &FlameGraph::setSearchValue);
+            connect(this, &FlameGraph::uiResetRequested, this, [this, searchInput] {
+                m_search.clear();
+                searchInput->clear();
+            });
+        },
+        this);
+
+    // use a QToolBar to automatically hide widgets in a menu that don't fit into the window
+    // the first 3 can stay as a widget since in their case the with is dominated by the number of tabs
```

Comment:
typo: with -> width

Reply:
also: why do you say "the first 3" - it looks like at least four widgets and three actions are added below - what does "the first 3" refer to?